### PR TITLE
Interactive/Cancelable slide left for details

### PR DESCRIPTION
**Pull request by michaelkirk**
This is much more tenable with the new perf improvements.

PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 1]
```diff
@@ -984,9 +988,14 @@ - (void)viewWillDisappear:(BOOL)animated
 
     [super viewWillDisappear:animated];
 
-    self.isViewVisible = NO;
-
     [self.inputToolbar viewWillDisappear:animated];
+}
+
+- (void)viewDidDisappear:(BOOL)animated
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: viewWillDisappear is called even if the transition away is cancelled. We should delay all possible teardown until viewDidDisappear.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 2]
```diff
@@ -162,6 +162,12 @@ - (void)commontInit
     UILongPressGestureRecognizer *longPress =
         [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(handleLongPressGesture:)];
     [self addGestureRecognizer:longPress];
+
+    PanDirectionGestureRecognizer *panGesture =
+        [[PanDirectionGestureRecognizer alloc] initWithDirection:PanDirectionHorizontal
+                                                          target:self
+                                                          action:@selector(handlePanGesture:)];
+    [self addGestureRecognizer:panGesture];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Should this GR hang on the cells or on the conversation collection view?  

* This won't work for pans that start in offers, errors, calls, unread indicator.
* This won't work for pans that start outside messages (on the left of outgoing, on the right of incoming).

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 1 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Maybe add a comment in `viewWillDisappear` noting that we shouldn't add logic there?

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 3]
```diff
@@ -4015,6 +4018,110 @@ - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
...
 }
 
+#pragma mark - swipe to show message details
+
+- (void)didPanWithGestureRecognizer:(UIPanGestureRecognizer *)gestureRecognizer
+                           viewItem:(ConversationViewItem *)conversationItem
+{
+    self.currentShowMessageDetailsPanGesture = gestureRecognizer;
+
+    const CGFloat leftTranslation = -1 * [gestureRecognizer translationInView:self.view].x;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Do we need to clamp this?  Could users reverse the gesture and pan back past the original arrangement?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 2 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Oh I see this is to show message metadata.  Please ignore.

**Comment by charlesmchen** [comment]
> This is much more tenable with the new perf improvements.

We gotta test this on iOS 8, 9, 10 and 11.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 1 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Good point. Done!

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m (hunk 3 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Not that I could reproduce, but I've now clamped it between 0-1.0 anyway.

**Pull request merged**